src/arch/arm/api/faults.c

```diff
@@ -37,7 +37,7 @@ word_t Arch_setMRs_fault(tcb_t *sender, tcb_t *receiver, word_t *receiveIPCBuffe
         if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
             word_t ipa, va;
             va = getRestartPC(sender);
-            ipa = (addressTranslateS1CPR(va) & ~MASK(PAGE_BITS)) | (va & MASK(PAGE_BITS));
+            ipa = (addressTranslateS1(va) & ~MASK(PAGE_BITS)) | (va & MASK(PAGE_BITS));
```

Comment:
Do we usually only invoke this on doFaultTransfer, which is invoked by doIPCTransfer to handle a fault? If so, I doubt we would hit this path for a VM.

Reply:
I can't really say much on that, need @kent-mcleod ... I'm fine to rewrite the commit message (will need it anyway), or even do a different code/proof update, but need kernel people to sort out what the desired behaviour is.

Reply:
Data faults and pre fetch faults still cause a fault to be delivered to the VM's fault handler, but as a VMFault type, not a VCPU fault type. So this path does get hit for a VCPU. Currently the Fault address is translated to IPA, and on arm_hyp the PC is translated to IPA, but on aarch64 the PC is left as VA.   arm_hyp has 2 definitions of `addressTranslateS1CPR`, the non-hyp version just returns its argument while the arm-hyp one does a translation.  The aarch64 architecture was only providing a single definition of the function, the non-hyp version and so was not translating properly.  This difference from arm-hyp, hadn't been noticed by us because our user level vm libraries don't to anything with this value other than print it out to console.  Most VMM fault handling code would obtain the PC from TCB_ReadRegisters and update it via TCB_WriteRegisters which consistently returns the VA on both architectures.

Reply:
That means I messed up yet another thing here, on aarch64 I made addressTranslateS1 always call `ats1e1r` whether hyp is on or not. From what you're saying addressTranslateS1 should have ifdefs to check for hyp, and return its argument in the no hyp case. Is that right?

Reply:
Yea.  I missed it was defined twice on arm-hyp yesterday.  A 2nd definition shouldn't really be needed as the function isn't called in non-hyp configs (because of the `if(0)` guard and other ifdefs), but it makes it consistent with arm32.

Reply:
I missed that part: sendIPC calls the function when dealing a VM fault. I think the interesting question here is what the VMM needs. On aarch64, for data abort, the fault IP is the guest VA IP and the fault address is the IPA. For instruction abort, the fault IP is the guest VA IP and the fault address is the IPA of the faulting instruction. When running a VM, usually printing out the GVA IP helps to match the fault with the guest kernel disassembly, and the VMM usually does not need to decode a faulting instruction since the FSR gives info about the fault. This looks fine for me. Another thing here is that for native seL4 threads, the translation here should not be needed given that there is S1 is disabled. Just my thoughts, could be incorrect.

Reply:
> That means I messed up yet another thing here, on aarch64 I made addressTranslateS1 always call `ats1e1r` whether hyp is on or not. From what you're saying addressTranslateS1 should have ifdefs to check for hyp, and return its argument in the no hyp case. Is that right? This mess-up might be the cause of some of the hardware failures.

ats1e1r might generate translation exceptions, but fortunately "Except as described in this section, these faults are not taken as an exception for the address translation instructions, but instead the PAR_EL1.FST field holds the fault status information. In these cases the PAR_EL1.PA field does not hold the output address of the translation."

Reply:
I'll go add it for consistency then. How would you improve the commit message to make sense to the target audience?

Reply:
Ok, added non-hyp `addressTranslateS1` on aarch64 for consistency.

Reply:
> For instruction abort, the fault IP is the guest VA IP and the fault address is the IPA of the faulting instruction. When running a VM, usually printing out the GVA IP helps to match the fault with the guest kernel disassembly, and the VMM usually does not need to decode a faulting instruction since the FSR gives info about the fault.

If you test it out at the moment, printing faults on the TK1 (arm-hyp) print an IPA for the faulting PC and printing faults on the TX2 (aarch64-hyp) print a VA for the faulting PC.  Using TCB_ReadRegister() for the PC gives you the VA on both and the Fault address is IPA on both.  So there's an argument that whatever way the PC is reported it should be the same for 32 and 64bit

Reply:
Another issue I vaguely remember here is that the translation assumes the fault-trigger's page tables. But it could be called from a fault handler's context, so the translation would be incorrect.

Reply:
> Another thing here is that for native seL4 threads, the translation here should not be needed given that there is S1 is disabled. Just my thoughts, could be incorrect.

`handleVMFault` does check `armHSVCPUActive` on aarch64 but not on aarch32.  Checking the armHSVCPUACtive state and avoiding the traversal is probably faster right?

Reply:
I am fine with both (32 hyp or 64 hyp) as long as I know where to get the addresses :-).  Adding a check means we might get different addresses based on the runtime behaviours?

Reply:
Well the check shouldn't change the outcome.  Doing a stage 1 translation for non-vcpu threads would just return the IPA because there is no stage 1 translation.

Reply:
> Well the check shouldn't change the outcome. Doing a stage 1 translation for non-vcpu threads would just return the IPA because there is no stage 1 translation.

1. We probably need to confirm that the AT instruction does not return an exception when the stage 1 is disabled. I guess this is true.
 
2. What I meant is that the function could be called when the VCPU is not running. For instance, the VCPU triggers an fault but the fault handler is not already waiting on that endpoint. The VCPU blocks, and the fault handler later receives the fault, and thus the function is call when the fault handler is running. AT instruction depends on the current context, in this the fault handler's context is used to translate the address, other than the VCPU's context.